sdk/storage/Azure.Storage.Common/src/Shared/AesGcm/AllowUnsafeBlocks/AeadCommon.cs

```diff
@@ -0,0 +1,130 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography;
+
+namespace Azure.Storage
+{
+    /// <summary>
+    /// Taken from
+    /// https://github.com/dotnet/runtime/blob/main/src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/AeadCommon.Windows.cs
+    ///
+    /// SHOULD NOT BE CHANGED WITHOUT COORDINATING WITH BCL TEAM
+    /// </summary>
+    internal static partial class AeadCommon
+    {
+        internal static unsafe void Encrypt(
+            Interop.BCrypt.SafeKeyHandle keyHandle,
+            ReadOnlySpan<byte> nonce,
+            ReadOnlySpan<byte> associatedData,
+            ReadOnlySpan<byte> plaintext,
+            Span<byte> ciphertext,
+            Span<byte> tag)
+        {
+            // bcrypt sometimes misbehaves when given nullptr buffers; ensure non-nullptr
+            fixed (byte* plaintextBytes = &GetNonNullPinnableReference(plaintext))
+            fixed (byte* nonceBytes = &GetNonNullPinnableReference(nonce))
+            fixed (byte* ciphertextBytes = &GetNonNullPinnableReference(ciphertext))
+            fixed (byte* tagBytes = &GetNonNullPinnableReference(tag))
+            fixed (byte* associatedDataBytes = &GetNonNullPinnableReference(associatedData))
+            {
+                var authInfo = Interop.BCrypt.BCRYPT_AUTHENTICATED_CIPHER_MODE_INFO.Create();
+                authInfo.pbNonce = nonceBytes;
+                authInfo.cbNonce = nonce.Length;
+                authInfo.pbTag = tagBytes;
+                authInfo.cbTag = tag.Length;
+                authInfo.pbAuthData = associatedDataBytes;
+                authInfo.cbAuthData = associatedData.Length;
+
+                Interop.BCrypt.NTSTATUS ntStatus = Interop.BCrypt.BCryptEncrypt(
+                    keyHandle,
+                   plaintextBytes,
+                    plaintext.Length,
+                    new IntPtr(&authInfo),
+                    null,
+                    0,
+                    ciphertextBytes,
+                    ciphertext.Length,
+                    out int ciphertextBytesWritten,
+                    0);
+
+                Debug.Assert(plaintext.Length == ciphertextBytesWritten);
+
+                if (ntStatus != Interop.BCrypt.NTSTATUS.STATUS_SUCCESS)
+                {
+                    throw Interop.BCrypt.CreateCryptographicException(ntStatus);
```

Comment:
Couldn't find an exception for this, I'm guessing if we get here it's unexpected?